src/dd/dd.rs

```diff
@@ -0,0 +1,749 @@
+                        }
+                    }
+                }
+            } else {
+                let mut to_skip = cf_guard.seek * cf_guard.out_block_size;
+                let mut buf = vec![0u8; BUF_SIZE];
+                while to_skip > 0 {
+                    match f.read(&mut buf[..cmp::min(to_skip, BUF_SIZE)]) {
+                        Ok(0) => break,
+                        Ok(m) => to_skip -= m,
+                        _ => err("cannot seek on output file".to_string()),
+                    }
+                }
+                if to_skip > 0 {
+                    while to_skip > 0 {
+                        match f.write(&zero_buf[..cmp::min(to_skip, BUF_SIZE)]) {
+                            Ok(0) => err("cannot seek on output file".to_string()),
+                            Ok(m) => to_skip -= m,
+                            _ => err("cannot seek on output file".to_string()),
+                        }
+                    }
+                }
+            }
+        }
+
+        // Would have terminated early if seek != 0, test precondition anyway
+        OutFileType::Stdout(_) => assert_eq!(cf_guard.seek, 0),
+    }
+    out
+}
+
+// , _sdone: chan::Sender<()>
+fn setup(cf: Arc<Mutex<Conf>>) -> (InpFileWrap, OutFileWrap) {
+    let mut cf_guard = cf.lock().unwrap();
+
+    let inp = if cf_guard.is_stdin {
+        Some(InpFileWrap {
+            f: InpFileType::Stdin(io::stdin()), // TODO: use .lock() if necessary
+            is_seekable: false,
+        })
+    } else {
+        let f = File::open(cf_guard.inp_name.as_ref() as &str);
+        if f.is_err() {
+            err("cannot open input file".to_string());
+            None
+        } else {
+            Some(InpFileWrap {
+                f: InpFileType::Reg(f.unwrap()),
+                is_seekable: true, // TODO: handle case when device isn't seekable
+            })
+        }
+    };
+    let mut inp = inp.unwrap();
+
+    let out = if cf_guard.is_stdout {
+        Some(OutFileWrap {
+            f: OutFileType::Stdout(io::stdout()), // TODO: use .lock() if necessary
+            is_seekable: false,
+        })
+    } else {
+        let f = File::create(cf_guard.out_name.as_ref() as &str);
+        if f.is_err() {
+            err("cannot open output file".to_string());
+            None
+        } else {
+            Some(OutFileWrap {
+                f: OutFileType::Reg(f.unwrap()),
+                is_seekable: true, // TODO: handle case when device isn't seekable
+            })
+        }
+    };
+    let mut out = out.unwrap();
+
+    drop(cf_guard);
+    inp = seek_inp(inp, Arc::clone(&cf));
+    out = seek_out(out, Arc::clone(&cf));
+
+    (inp, out)
+}
+
+fn copy(
+    mut inp: InpFileWrap,
+    mut out: OutFileWrap,
+    cf_ptr: Arc<Mutex<Conf>>,
+    _sdone: chan::Sender<()>,
+) -> result::Result<(), String> {
+    {
+        let mut cf = cf_ptr.lock().unwrap();
```

Comment:
In order to prevent the signal handling thread from blocking until the end of the program, you may need to reacquire the lock on every loop iteration.  If you do so it might be better to just avoid using `chan_select!` and spawn a channel using `chan::async()`, which you then pass to `notify_on()` for each of the signals.  You would then have to check at the bottom of the `for` loop here if a signal was passed and react accordingly.